verification/io/io-spec.gobra

```diff
@@ -38,75 +38,191 @@ type Place int
 pred token(t Place)
 
 ghost
+decreases
 pure func undefined() IO_dp3s_state_local
 
-pred dp3s_iospec_bio3s_enter(s IO_dp3s_state_local, t Place)
-/*
-∀⋆ (λv. case v of
-Internal_val1 pkt recvif newpkt nextif ⇒
-if pkt ∈ ibuf s (Some recvif) ∧ (∃currseg traversedseg.
-(∃hf1 fut. dp2_enter_guard pkt currseg traversedseg asid hf1 recvif fut) ∧
-dp3s_forward (Packet2 traversedseg (LeftSeg pkt) (MidSeg pkt) (RightSeg pkt)) newpkt
-nextif)
-then ExT (λT'. ExV (λX. CBio (IN bio3s_enter) t (Internal_val1 pkt recvif newpkt nextif) X T'
-⋆ dp3s_iospec_ordered (dp3s_add_obuf s nextif newpkt) T')) else Bool True | _ ⇒ Bool True)
-*/
-
-pred dp3s_iospec_bio3s_xover_up2down(s IO_dp3s_state_local, t Place)
-/*
-∀⋆ (λv. case v of
-Internal_val1 pkt recvif newpkt nextif ⇒
-if ∃currseg. ¬ ConsDir currseg ∧ (∃nextseg. ConsDir nextseg ∧
-(∃traversedseg intermediatepkt hf1 hf2.
-xover_up2down2_link_type asid hf1 hf2 ∧
-(∃nextfut. dp3s_xover_common s pkt currseg nextseg traversedseg intermediatepkt hf1
-hf2 nextfut recvif newpkt nextif)))
-then ExT (λT'. ExV (λX. CBio (IN bio3s_xover_up2down) t (Internal_val1 pkt recvif newpkt
-nextif) X T'
-)
-⋆ dp3s_iospec_ordered (dp3s_add_obuf s nextif newpkt) T')) else Bool True
-| _ ⇒ Bool True)
-*/
-
-pred dp3s_iospec_bio3s_xover_core(s IO_dp3s_state_local, t Place)
-/* s t ≡ ∀⋆ (λv. case v of
-Internal_val1 pkt recvif newpkt nextif ⇒
-if asid ∈ core ∧ (∃currseg nextseg.
-ConsDir nextseg = ConsDir currseg ∧ (∃traversedseg intermediatepkt hf1 hf2.
-xover_core2_link_type hf1 hf2 asid (ConsDir currseg) ∧
-(∃nextfut. dp3s_xover_common s pkt currseg nextseg traversedseg intermediatepkt hf1 hf2 nextfut recvif newpkt nextif)))
-then ExT (λT'. ExV (λX. CBio (IN bio3s_xover_core) t (Internal_val1 pkt recvif newpkt nextif) X
-T' * dp3s_iospec_ordered (dp3s_add_obuf s nextif newpkt) T')) else Bool True
-| _ ⇒ Bool True)
-*/
-
-pred dp3s_iospec_bio3s_exit(s IO_dp3s_state_local, t Place)
-/* s t ≡ ∀⋆ (λv. case v of
-Internal_val2 pkt newpkt nextif ⇒
-if pkt ∈ ibuf s None ∧ dp3s_forward_ext pkt newpkt nextif
-then ExT (λT'. ExV (λX. CBio (IN bio3s_exit) t (Internal_val2 pkt newpkt nextif) X T' ⋆ dp3s_iospec_ordered (dp3s_add_obuf s (Some nextif) newpkt) T'))
-else Bool True | _ ⇒ Bool True)
-*/
-
-pred CBioIO_bio3s_send(t Place)
+pred CBio_IN_bio3s_enter(t Place, v IO_val)
 
 ghost
-requires CBioIO_bio3s_send(t)
-pure func dp3s_iospec_bio3s_send_T(s IO_dp3s_state_local, t Place) Place
+requires CBio_IN_bio3s_enter(t, v)
+decreases
+pure func CBio_IN_bio3s_enter_T(s IO_dp3s_state_local, t Place, v IO_val) Place
 
+
+// This corresponds to the condition of the if statement in the io-spec case for enter
+ghost
+decreases
+pure func dp3s_iospec_bio3s_enter_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return some(v.IO_Internal_val1_2) in domain(s.ibuf) ==>
+		(let ibuf_set := s.ibuf[some(v.IO_Internal_val1_2)] in (v.IO_Internal_val1_1 in ibuf_set)) &&
+		(exists currseg IO_seg3, traversedseg IO_seg3 ::
```

Comment:
this existencial looks very unnecessary. If we inline the definition of `dp2_enter_guard` here, we realize that `currseg` must be `v.IO_Internal_val1_1.CurrSeg` and `traversedseg`, although more complicated, is indirectly defined relationally in function `inc_seg2`.

One thing we can do is to create a computational, instead of relational version of these definitions to avoid the existencial. For example, we could create a version of `inc_seg2` like the following
```
ghost
decreases
// notice the name difference
ensures inc_seg2(currseg, traversedseg, head(currseg.Future), tail(curreseg.Future)) // notice this postcondition
pure func inc_seg2_comp(currseg IO_seg2) (traversedseg IO_seg2) {
	return IO_seg2_ {
        Future: tail(currseg.Future)
        Past: tail(currseg.Past)  
        History: seq[IO_ahi]{head(currseg.Future.)Toab()} ++ currseg.History
}
```
and use it instead of our current definition:
```
ghost
decreases
pure func inc_seg2(currseg, traversedseg IO_seg2, hf1 IO_HF, fut seq[IO_HF]) bool {
	return currseg.Future === seq[IO_HF]{hf1} ++ fut &&
		traversedseg.Future === fut &&
		traversedseg.Past === seq[IO_HF]{hf1} ++ currseg.Past &&
		traversedseg.History === seq[IO_ahi]{hf1.Toab()} ++ currseg.History
}
```

Reply:
We can also do this in a separate PR